src/db/spanner/batch.rs

```diff
@@ -269,22 +253,74 @@ pub async fn do_append_async(
     //  ("<fxa_uid>", "<fxa_kid>", 101, "ba1", "bso2", NULL, "payload2", NULL)]
     // https://cloud.google.com/spanner/docs/structs#creating_struct_objects
     let mut running_size: usize = 0;
-    let rows: Vec<_> = bsos
-        .into_iter()
-        .map(|bso| {
-            let sortindex = bso
-                .sortindex
-                .map(|sortindex| as_value(sortindex.to_string()))
-                .unwrap_or_else(null_value);
-            let payload = bso.payload.map(as_value).unwrap_or_else(null_value);
-            if payload != null_value() {
-                running_size += payload.get_string_value().len();
-            }
-            let ttl = bso
-                .ttl
-                .map(|ttl| as_value(ttl.to_string()))
-                .unwrap_or_else(null_value);
 
+    // problem: Append may try to insert a duplicate record into the batch_bsos table.
+    // this is because spanner doesn't do upserts. at all. Batch_bso is a temp table and
```

Comment:
It does support a form of upsert but only via mutations (mentioned [here](https://github.com/mozilla-services/syncstorage-rs/issues/618#issuecomment-680227710))